pkg/backends/calico/client.go

```diff
@@ -809,30 +815,30 @@ func (c *client) OnUpdates(updates []api.Update) {
 			log.Info("Starting route generator due to service advertisement update")
 			var err error
 			if c.rg, err = NewRouteGenerator(c); err != nil {
-				log.WithError(err).Error("Failed to start route generator, unable to advertise services")
+				log.WithError(err).Error("Failed to start route generator, unable to advertise node-specific service routes")
 				c.rg = nil
 			} else {
 				c.rg.Start()
 			}
 		}
 
-		if c.rg != nil {
-			// Update CIDRs. In v1 format, they are a single comma-separate string. If the string isn't empty,
-			// split on the comma and pass a list of strings to the route generator.
-			// An empty string indicates a withdrawal of that set of service IPs.
-			var externalIPs []string
-			if len(c.cache["/calico/bgp/v1/global/svc_external_ips"]) > 0 {
-				externalIPs = strings.Split(c.cache["/calico/bgp/v1/global/svc_external_ips"], ",")
-			}
-			var clusterCIDRs []string
-			if len(c.cache["/calico/bgp/v1/global/svc_cluster_ips"]) > 0 {
-				clusterCIDRs = strings.Split(c.cache["/calico/bgp/v1/global/svc_cluster_ips"], ",")
-			}
+		// Update external IP CIDRs. In v1 format, they are a single comma-separated
+		// string. If the string isn't empty, split on the comma and pass a list of strings
+		// to the route generator.  An empty string indicates a withdrawal of that set of
+		// service IPs.
+		if len(c.cache["/calico/bgp/v1/global/svc_external_ips"]) > 0 {
+			c.onExternalIPsUpdate(strings.Split(c.cache["/calico/bgp/v1/global/svc_external_ips"], ","))
+		}
+
+		// Same for cluster CIDRs.
+		if len(c.cache["/calico/bgp/v1/global/svc_cluster_ips"]) > 0 {
+			c.onClusterIPsUpdate(strings.Split(c.cache["/calico/bgp/v1/global/svc_cluster_ips"], ","))
+		}
 
-			// Run these as separate goroutines so that we don't block. We're already holding the lock,
-			// and these calls may attempt to take the lock in order to update the cache.
-			go c.rg.onExternalIPsUpdate(externalIPs)
-			go c.rg.onClusterIPsUpdate(clusterCIDRs)
+		if c.rg != nil {
+			// Ask the route generator to recheck and advertise or withdraw
+			// node-specific routes.
+			go c.rg.resyncKnownRoutes()
```

Comment:
I don't think this is sound; if you get two goroutines in flight then they could both block on the `rg.Lock()` call in `setRoutesForSvc` and get reordered.

Reply:
I don't think reordering matters for what resyncKnownRoutes does, though.

Reply:
Anyway, I believe I've now addressed this with https://github.com/projectcalico/confd/pull/323/commits/8bb66962fca82634e32efea3722c35902cb17058

Reply:
Here's the problem that I think could happen:

- goroutine A started, reads `Service{Type: "ExternalName"}`
- service gets changes form ExternalName to ClusterIP
- goroutine B started, reads `Service{Type: "ClusterIP"}`
- goroutine B scheduled, sends update `Service{Type: "ClusterIP"}`
- goroutine A schedules, sends update `Service{Type: "ExternalName"}`

End result: we think the service is of type `ExternalName` instead of type `ClusterIP` and we withdraw its route.

Reply:
Agreed.  But I've addressed that already now with https://github.com/projectcalico/confd/commit/8bb66962fca82634e32efea3722c35902cb17058